File: src/main/java/nz/govt/natlib/dashboard/common/auth/Sessions.java
package nz.govt.natlib.dashboard.common.auth;

import org.springframework.stereotype.Component;

import java.util.Date;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Basically a look-up table, containing a map of session IDs to session info structs
 */
@Component
public class Sessions {

    private final ConcurrentHashMap<String, SessionInfo> sessionMap = new ConcurrentHashMap<>();

    /**
     * Add session and remove any expired sessions while we're at it
     */
    public void addSession(String id, String username, String role, long expireInterval) {
        if (sessionMap.containsKey(id)) {
            // Can't happen
            throw new RuntimeException(String.format("Session id %s already exists", id));
        }
        for (String key : sessionMap.keySet()) {
            if (sessionMap.get(key).expired()) {
                removeSession(id);
            }
        }
        sessionMap.put(id, new SessionInfo(username, role, expireInterval));
    }

    public void removeSession(String id) {
        sessionMap.remove(id);
    }

    /**
     * Get the username for this session if it's still valid and, if so, update the timestamp of last access
     */
    public String getUsername(String id) throws InvalidSessionException {
        if (!sessionMap.containsKey(id) || sessionMap.get(id).expired()) {
            sessionMap.remove(id);
            throw new InvalidSessionException(id);
        }
        sessionMap.get(id).touch();
        return sessionMap.get(id).username;
    }

    /**
     * Get the role for this session if it's still valid and, if so, update the timestamp of last access
     */
    public String getRole(String id) throws InvalidSessionException {
        if (!sessionMap.containsKey(id) || sessionMap.get(id).expired()) {
            sessionMap.remove(id);
            throw new InvalidSessionException(id);
        }
        sessionMap.get(id).touch();
        return sessionMap.get(id).role;
    }

    public static class InvalidSessionException extends Exception {
        public InvalidSessionException(String id) {
            super(String.format("Session with %s has expired", id));
        }
    }


    // Struct used as the value in the kv pair representing a session
    static class SessionInfo {
        Date modified;
        String username;
        String role;
        long expireInterval;

        public SessionInfo(String username, String role, long expireInterval) {
            this.modified = new Date();
            this.username = username;
            this.role = role;
            this.expireInterval = expireInterval;
        }

        boolean expired() {
            return System.currentTimeMillis() - modified.getTime() > expireInterval;
        }

        void touch() {
            modified = new Date();
        }
    }
}

File: server/app/auth/sessions.py
import enum
import logging
from dataclasses import dataclass
from typing import Dict
import time


class RoleType(enum.Enum):
    BOOTSTRAP = "bootstrap"
    ADMIN = "admin"
    NORMAL = "normal"


@dataclass
class SessionInfo:
    modified: float
    username: str
    role: RoleType
    expire_interval: int

    def expired(self):
        return time.time() - self.modified > self.expire_interval

    def touch(self):
        self.modified = time.time()


class SessionManager:
    def __init__(self, expire_interval):
        self.expire_interval = expire_interval
        self.session_map: Dict[str:SessionInfo] = {}

    def add_session(self, token: str, username: str, role: RoleType):
        if token in self.session_map:
            raise RuntimeError(f"Session {token} already exists")

        keys = list(self.session_map.keys())
        for key in keys:
            s: SessionInfo = self.session_map.get(key)
            if s.expired():
                self.session_map.pop(key)
                logging.info(f"{s.username} {key} is expired")

        s = SessionInfo(
            modified=time.time(),
            username=username,
            role=role,
            expire_interval=self.expire_interval,
        )
        self.session_map[token] = s

    def remove_session(self, token):
        if token in self.session_map:
            self.session_map.pop(token)

    def get_username(self, token):
        if token not in self.session_map:
            raise RuntimeError(f"Invalid session: {token}")
        s: SessionInfo = self.session_map.get(token)
        if not s.expired():
            s.touch()
        return s.username

    def get_role(self, token):
        if token not in self.session_map:
            raise RuntimeError(f"Invalid session: {token}")
        s: SessionInfo = self.session_map.get(token)
        if not s.expired():
            s.touch()
        return s.role

    def is_valid(self, token):
        if token not in self.session_map:
            return False

        s: SessionInfo = self.session_map.get(token)
        if s.expired():
            return False

        s.touch()
        return True
